dev/RadioButtons/RadioButtons.cpp

```diff
@@ -7,136 +7,634 @@
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if (currentIndex && currentIndex.GetSize() == 1)
+    {
+        auto const index = currentIndex.GetAt(0);
+        if (repeater)
+        {
+            if (auto const item = repeater.TryGetElement(index))
+            {
+                if (auto const itemAsToggleButton = item.try_as<winrt::ToggleButton>())
+                {
+                    itemAsToggleButton.IsChecked(true);
+                }
+            }
+        }
+        UpdateSelectionDPs(index);
+    }
+    else
+    {
+        UpdateSelectionDPs(-1);
     }
 }
 
-void RadioButtons::OnListViewKeyUp(const winrt::IInspectable& sender, const winrt::KeyRoutedEventArgs& args)
+void RadioButtons::UpdateSelectionDPs(const int newIndex)
 {
-    if (args.Key() == winrt::VirtualKey::Control)
+    if (auto const repeater = m_repeater.get())
     {
-        m_isControlDown = false;
+        auto const oldSelectedItemsRepeaterIndex = [selectedItem = SelectedItem(), repeater]()
+        {
+            if (selectedItem)
+            {
+                return repeater.GetElementIndex(selectedItem.try_as<winrt::UIElement>());
+            }
+            return -1;
+        }();
+        if (SelectedIndex() != newIndex || oldSelectedItemsRepeaterIndex != newIndex)
+        {
+            SelectedIndex(newIndex);
+            auto const eventArgs = [this, newIndex]()
+            {
+                auto const previousSelectedItem = SelectedItem();
+                if (auto const repeater = m_repeater.get())
+                {
+                    if (auto const newSelectedItem = repeater.TryGetElement(newIndex))
+                    {
+                        SelectedItem(newSelectedItem);
+                        return winrt::SelectionChangedEventArgs({ previousSelectedItem }, { newSelectedItem });
+                    }
+                }
+                SelectedItem(nullptr);
+                return winrt::SelectionChangedEventArgs({ previousSelectedItem }, { nullptr });
+            }();
+            m_selectionChangedEventSource(*this, eventArgs);
+        }
+    }
+}
+
+void RadioButtons::OnChildChecked(const winrt::IInspectable& sender, const winrt::RoutedEventArgs&)
+{
+    if (auto const repeater = m_repeater.get())
+    {
+        if (auto const senderAsUIE = sender.as<winrt::UIElement>())
+        {
+            m_selectionModel.Select(repeater.GetElementIndex(senderAsUIE));
+        }
     }
 }
 
-bool RadioButtons::MoveSelection(int direction)
+void RadioButtons::OnChildUnchecked(const winrt::IInspectable& sender, const winrt::RoutedEventArgs&)
 {
-    bool found = false;
+    if (auto const repeater = m_repeater.get())
+    {
+        if (auto const senderAsUIE = sender.as<winrt::UIElement>())
+        {
+            m_selectionModel.Deselect(repeater.GetElementIndex(senderAsUIE));
+        }
+    }
+}
+
+bool RadioButtons::MoveFocusNext()
```

Comment:
Looks like a lot of the code is Focus and Keyboard navigation related. We should investigate it this is a common pattern and see if we can make this reusable.

Reply:
I think if we could remove the radio Button key down handler this would almost all go away, which makes me think that this probably isn't very useful outside of this instance. But if it is I totally agree. 

---
In reply to: [347433351](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347433351) [](ancestors = 347433351)

Reply:
xyfocus eliminated a lot of it. Still might be interesting to factor this out though.

---
In reply to: [347545732](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347545732) [](ancestors = 347545732,347433351)